core/src/worker/activities.rs

```diff
@@ -509,6 +527,53 @@ where
                                     tt,
                                     ActivityCancelReason::WorkerShutdown,
                                 ));
+                            } else {
+                                // Fire off task to keep track of local timeouts. We do this so that
+                                // activities can still get cleaned up even if the user isn't
+                                // heartbeating.
+                                let local_timeout_buffer = self.local_timeout_buffer;
+                                let sched = task
+                                    .resp
+                                    .schedule_to_close_timeout
+                                    .clone()
+                                    .and_then(|to| {
+                                        // Must filter out zero before any subtraction math here
+                                        if to.seconds == 0 && to.nanos == 0 {
+                                            None
+                                        } else {
+                                            Some(to)
+                                        }
+                                    })
+                                    .zip(task.resp.scheduled_time.clone())
+                                    .and_then(|(to, st)| {
+                                        let now = SystemTime::now();
+                                        let sched_time = SystemTime::try_from(st).unwrap_or(now);
+                                        let already_elapsed =
+                                            now.duration_since(sched_time).unwrap_or_default();
+                                        Duration::try_from(to).map(|to| to - already_elapsed).ok()
+                                    });
+                                let timeout_at = [
+                                    task.resp.heartbeat_timeout.clone(),
+                                    task.resp.start_to_close_timeout.clone(),
+                                ]
+                                .into_iter()
+                                .flatten()
+                                .map(Duration::try_from)
+                                .filter_map(Result::ok)
+                                .chain(sched)
+                                .filter(|d| !d.is_zero())
+                                .min();
```

Comment:
This is the rustiest piece of code I've ever seen 😆

Reply:
:crab:

Reply:
Got a 10-chainer at https://github.com/temporalio/sdk-python/blob/2c15ed30159242d7f93534e8263e221de1959e1e/temporalio/bridge/src/metric.rs#L273-L283